Mathlib/RingTheory/PolynomialLaw/Basic.lean

```diff
@@ -0,0 +1,265 @@
+  ext S; simp only [smul_def, _root_.zero_smul, zero_def, Pi.zero_apply]
+
+theorem one_smul : (1 : R) • f = f := by
+  ext S; simp only [smul_def, Pi.smul_apply, _root_.one_smul]
+
+instance : MulAction R (M →ₚ[R] N) where
+  one_smul       := one_smul
+  mul_smul a b f := by ext; simp only [smul_def, mul_smul]
+
+instance : AddCommMonoid (M →ₚ[R] N) where
+  add_assoc f g h := by ext; simp only [add_def, add_assoc]
+  zero_add f      := by ext; simp only [add_def, zero_add, zero_def]
+  add_zero f      := by ext; simp only [add_def, add_zero, zero_def]
+  nsmul n f       := (n : R) • f
+  nsmul_zero f    := by simp only [Nat.cast_zero, zero_smul f]
+  nsmul_succ n f  := by simp only [Nat.cast_add, Nat.cast_one, add_smul, one_smul]
+  add_comm f g    := by ext; simp only [add_def, add_comm]
+
+instance : Module R (M →ₚ[R] N) where
+  smul_zero a    := rfl
+  smul_add a f g := by ext; simp only [smul_def, add_def, smul_add]
+  add_smul       := add_smul
+  zero_smul      := zero_smul
+
+end CommSemiring
+
+section CommRing
+variable {R : Type u} [CommRing R]
+  {M : Type*} [AddCommGroup M] [Module R M] {N : Type*} [AddCommGroup N] [Module R N]
+  (f : M →ₚ[R] N)
+
+/-- The opposite of a polynomial law -/
+noncomputable def neg : M →ₚ[R] N where
+  toFun' S _ _ := (-1 : R) • f.toFun' S
+  isCompat' φ  := by
+    ext
+    simp only [map_smul, Function.comp_apply, Pi.neg_apply, map_neg, isCompat_apply', Pi.smul_apply]
+
+instance : Neg (M →ₚ[R] N) := ⟨neg⟩
+
+@[simp]
+theorem neg_def (S : Type u) [CommSemiring S] [Algebra R S] :
+    (-f).toFun' S = (-1 : R) • f.toFun' S := rfl
+
+instance : AddCommGroup (M →ₚ[R] N) where
+  zsmul n f := (n : R) • f
+  zsmul_zero' f   := by simp only [Int.cast_zero, zero_smul]
+  zsmul_succ' n f := by simp only [Int.ofNat_eq_coe, Nat.cast_succ, Int.cast_add, Int.cast_natCast,
+    Int.cast_one, add_smul, _root_.one_smul]
+  zsmul_neg' n f  := by
+    ext S _ _ m
+    rw [neg_def]
+    simp only [Int.cast_negSucc, Nat.cast_add, Nat.cast_one, neg_add_rev, _root_.add_smul,
+      add_def_apply, smul_def_apply, Nat.succ_eq_add_one, Int.cast_add, Int.cast_natCast,
+      Int.cast_one, _root_.one_smul, add_def, smul_def, Pi.smul_apply, Pi.add_apply, smul_add,
+      smul_smul, neg_mul, one_mul]
+    rw [add_comm]
+  neg_add_cancel f  := by
+    ext S _ _ m
+    simp only [add_def_apply, neg_def, Pi.smul_apply, zero_def, Pi.zero_apply]
+    nth_rewrite 2 [← _root_.one_smul (M := R) (b := f.toFun' S m)]
+    rw [← _root_.add_smul]
+    simp only [neg_add_cancel, _root_.zero_smul]
+  add_comm f g    := by ext; simp only [add_def, add_comm]
+
+end CommRing
+
+end Module
+
+section ground
+
+variable {R : Type u} [CommSemiring R] {M : Type*} [AddCommMonoid M] [Module R M]
+  {N : Type*} [AddCommMonoid N] [Module R N]
+variable (f : M →ₚ[R] N)
+
+/-- The map `M → N` associated with a `f : M →ₚ[R] N` (essentially, `f.toFun' R`) -/
+def ground : M → N := (TensorProduct.lid R N) ∘ (f.toFun' R) ∘ (TensorProduct.lid R M).symm
```

Comment:
I would suggest doing this:
```lean
instance : CoeFun (M →ₚ[R] N) (fun _ ↦ M → N) where
  coe := ground
```
Given `f : M →ₚ[R] N` and `m : M`, this allows the syntax `f m : N`.